### PR TITLE
Use correct check for slider path extension

### LINKED ISSUE #24798: Slider length calculation does not match Stable for some edge cases
Copied over from https://github.com/ppy/osu/issues/24708#issuecomment-1717192869

***

[List of approved maps with incorrect `max_combo`](https://docs.google.com/spreadsheets/d/19TFgR_ZlKEWeqCnKoXJlf48Scql0QDl3YAA5ppw9MNg), majority can be ignored as they abuse things like infinite bpm/negative slider lengths/etc.

#24708 covers beatmaps [`1855`](https://osu.ppy.sh/beatmapsets/268#osu/1855) and [`4182386`](https://osu.ppy.sh/beatmapsets/2007300#osu/4182386). But for [`42196`](https://osu.ppy.sh/beatmapsets/10804#osu/42196), the slider length is off by a huge amount. This is because the length specified in the [.osu file](https://osu.ppy.sh/osu/42196) is different than the length calculated using the path points. For Stable, the calculated length is `222.004464186728` while the length in the beatmap file is `330`. Stable attempts to resolve this by removing the "excess", which happens to add the missing amount back when the "excess" is negative. After doing this, the calculated length becomes `330.00000862032175`.

For Lazer, the recalculation for the last path is skipped when the last two control points are equal and when the expected length is greater than the calculated length. Relevant Lazer code: https://github.com/ppy/osu/blob/f90f2491c349c58f5a038ab83e9934ca4b483787/osu.Game/Rulesets/Objects/SliderPath.cs#L331-L336
Meanwhile in Stable, the only scenario where recalculation is skipped is when the last two control points are equal and when the length of the final path is greater than the "excess" (calculated length - expected length). Applying this logic to Lazer results in the same slider length, although I'm unsure if this will have any unintended consequences (also not sure if the `0.0001` is needed):
```diff
diff --git a/osu.Game/Rulesets/Objects/SliderPath.cs b/osu.Game/Rulesets/Objects/SliderPath.cs
index 34113285a4..b4507f8d6f 100644
--- a/osu.Game/Rulesets/Objects/SliderPath.cs
+++ b/osu.Game/Rulesets/Objects/SliderPath.cs
@@ -328,16 +328,16 @@ private void calculateLength()

             if (ExpectedDistance.Value is double expectedDistance && calculatedLength != expectedDistance)
             {
+                int pathEndIndex = calculatedPath.Count - 1;
+
                 // In osu-stable, if the last two control points of a slider are equal,  extension is not performed.
-                if (ControlPoints.Count >= 2 && ControlPoints[^1].Position == ControlPoints[^2].Position && expectedDistance > calculatedLength)
+                if (ControlPoints.Count >= 2 && ControlPoints[^1].Position == ControlPoints[^2].Position && calculatedPath[pathEndIndex].Length > calculatedLength - expectedDistance + 0.0001)
                 {
                     cumulativeLength.Add(calculatedLength);
                     return;
                 }
 
                 // The last length is always incorrect
                 cumulativeLength.RemoveAt(cumulativeLength.Count - 1);
 
-                int pathEndIndex = calculatedPath.Count - 1;
-
                 if (calculatedLength > expectedDistance)
```
***
## Stable
screenshot4476
## Lazer
osu_2023-09-13_03-06-20
## Lazer with diff applied
osu_2023-09-13_03-33-00

### PR DESCRIPTION
**Pull request by Magnus-Cosmos**
Currently, Lazer uses the last two control points of a slider to check if it needs an extension: https://github.com/ppy/osu/blob/163b6df357a620dbee6ef0144f129e5ed9be682f/osu.Game/Rulesets/Objects/SliderPath.cs#L331-L336 However Stable uses the points of the last calculated path. In most cases this makes no difference, but for [`42196`](https://osu.ppy.sh/beatmapsets/10804#osu/42196), this discrepancy causes the sliders at `46060ms` and `123348ms` to skip path extension in Lazer, making the sliders shorter compared to Stable.

Closes #24798.

***

While fixing this, I noticed that there are also some other slider length discrepancies:
-  In the test case provided by this PR, the sliders have 64 calculated paths in Stable but 32 in Lazer. For some reason Lazer converts the first segment of the slider into a Linear path, but uses Bezier for the second segment.
- Lazer calculates slider curves using points relative to the start position. This causes the floating point precision loss to be different during calculations compared to Stable, leading to slightly different path lengths.

I believe these issues can be ignored or addressed later, as I have yet to encounter any edge cases where these make a noticeable difference.

### CODE REVIEW

**Comment by Magnus-Cosmos** [comment]
Currently failing this edge case due to a discrepancy between Stable and Lazer path calculation:
https://github.com/ppy/osu/blob/163b6df357a620dbee6ef0144f129e5ed9be682f/osu.Game.Tests/Beatmaps/Formats/LegacyBeatmapDecoderTest.cs#L870

Slider used in the test, from #15303:
`261,171,25305,6,0,B|262:171|262:171|262:171,1,2,8|0,0:0|0:0,0:0:0:0:`

Relevant methods:
https://github.com/ppy/osu/blob/163b6df357a620dbee6ef0144f129e5ed9be682f/osu.Game/Rulesets/Objects/SliderPath.cs#L242 https://github.com/ppy/osu/blob/163b6df357a620dbee6ef0144f129e5ed9be682f/osu.Game/Rulesets/Objects/SliderPath.cs#L282
Results for the methods:
| subControlPoints    | type   | output             |
|---------------------|--------|--------------------|
| [{(0,0)}]           | Bezier | [{(0,0)}, {(0,0)}] |
| [{(0,0)}, {(1,0)}]  | Bezier | [{(0,0)}, {(1,0)}] |
| [{(1,0)}, {(1,0)}]  | Bezier | [{(1,0)}, {(1,0)}] |

| calculatePath()    | expected                    |
|--------------------|-----------------------------|
| [{(0,0)}, {(1,0)}] | [{(0,0)}, {(1,0)}, {(1,0)}] |

In `calculatePath()`, when adding the calculated sub paths to `calculatedPath`, for every path of the sub paths, it checks if the sub path is equal to the last path added. The purpose of this is to get rid of the points shared between each path.
https://github.com/ppy/osu/blob/163b6df357a620dbee6ef0144f129e5ed9be682f/osu.Game/Rulesets/Objects/SliderPath.cs#L267 This works for most scenarios, since having 3 consecutive curve points with the same position shouldn't happen normally. However, when it does happen, Lazer ends up skipping the 3rd sub path completely, resulting in 2 path points instead of 3 like in Stable.

There's 2 simple ways to fix this, either exclude the starting point of each sub path in `calculateSubPath()`'s output, or exclude it in `calculatePath()` when merging the sub paths. I've opted to do the latter since it's simpler.

**Comment by bdach** [comment]
Doing a SR sheet for this overnight as a preliminary step. Haven't reviewed the logic closely yet, just want to see how much changes.

**Comment by bdach** [comment]
Seems like diffcalc... doesn't care again? https://docs.google.com/spreadsheets/d/15Vgipe7vv7sK-4SUDwL1kj0WFE6g1v2VsPaXlIcqEoM/edit#gid=11450370

I'll have to do a manual check against stable source then...

**Comment by bdach** [comment]
Ok, upon comparison with stable source I _believe_ this is correct, although I was almost convinced it wasn't for a while...

Relevant snippet:

```csharp
                while (path.Count > 0)
                {
                    Line lastLine = path[path.Count - 1];
                    float lastLineLength = Vector2.Distance(lastLine.p1, lastLine.p2);

                    if (lastLineLength > excess + MIN_SEGMENT_LENGTH)
                    {
                        if (lastLine.p2 != lastLine.p1)
                        {
                            lastLine.p2 = lastLine.p1 +
                                      Vector2.Normalize(lastLine.p2 - lastLine.p1) * (lastLine.rho - (float)excess);
                        }
                        break;
                    }

                    path.Remove(lastLine);
                    excess -= lastLineLength;
                }
```

(source: https://github.com/peppy/osu-stable-reference/blob/1531237b63392e82c003c712faa028406073aa8f/osu!/GameplayElements/HitObjects/Osu/SliderOsu.cs#L712-L729).

And yes, `path` is indeed the path post-approximation rather than the list of path control points.

@Magnus-Cosmos Since you appear to _somehow_ have access to stable source (judging by the debugger screenshots from https://github.com/ppy/osu/issues/24708), it would be appreciated if you provided specific information as to which parts of stable specifically these types of changes are based on. It would make review much faster in the future if you should wish to continue submitting such PRs.

File: osu.Game/Rulesets/Objects/SliderPath.cs [hunk 1]
```diff
@@ -262,9 +262,16 @@ private void calculatePath()
                 var segmentVertices = vertices.AsSpan().Slice(start, i - start + 1);
                 var segmentType = ControlPoints[start].Type ?? PathType.Linear;
 
-                foreach (Vector2 t in calculateSubPath(segmentVertices, segmentType))
+                // No need to calculate path when there is only 1 vertex
+                if (segmentVertices.Length == 1)
+                    calculatedPath.Add(segmentVertices[0]);
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/SliderPath.cs:267]: This appears to be a mostly unrelated optimisation, but I can't see it being wrong and sheets are silend, so I guess I'll let it be...

**Comment by bdach** [comment]
Did some further checking against maps mentioned in #15109, confirmed that this change doesn't regress either case mentioned there.

I can't find any more fault with this change so I'm just gonna take a somewhat calculated risk on it and hope for the best...

**Review by bdach — APPROVED** [review]
